Add param subseq_isconstant to floss

NimaSarajpoor opened the pull request:
This PR adds subseq_isconstant param to `floss` module

NimaSarajpoor reviewed (commented):
@seanlaw 
I think `floss` is almost ready. Please let me know what you think about the comments I provided here.

NimaSarajpoor commented on stumpy/floss.py:
```diff
@@ -476,6 +494,16 @@ def __init__(
+            )
+            raise ValueError(msg)
+        else:
+            self._isconstant_func = T_subseq_isconstant
```
I think your proposal is clearer as the suggested name follows the structure of the original `T_subseq_isconstant`. I will revise the name of variables.

NimaSarajpoor commented on stumpy/floss.py:
```diff
@@ -347,6 +347,14 @@ class floss:
     normalize : bool, default True
         When set to `True`, this z-normalizes subsequences prior to computing distances
 
```
Sure. I will fix it.

NimaSarajpoor commented on stumpy/floss.py:
```diff
@@ -476,6 +494,16 @@ def __init__(
+        if not callable(T_subseq_isconstant):  # pragma: no cover
+            msg = (
+                "The parameter `T_subseq_isconstant` must be "
+                + "callabel. For details, see the docstring."
```
Thank you for catching the typo. I will fix it and revies the message.

NimaSarajpoor commented on stumpy/floss.py:
```diff
@@ -476,6 +494,16 @@ def __init__(
+                + "callabel. For details, see the docstring."
+            )
+            raise ValueError(msg)
+        else:
```
Right! I will remove it!

NimaSarajpoor commented on tests/test_floss.py:
```diff
@@ -411,3 +419,85 @@ def test_aamp_floss_inf_nan(substitute, substitution_locations):
+    def custom_func(a, w):
+        _, Σ_a = core.compute_mean_std(a, w)
+
+        return Σ_a < 0.5
```
The function `isconstant_func_stddev_threshold(T, m, ...)` decides a subsequence is constant or not based on its standard deviation and a threshold, denoted as `stddev_threshold` in the code. However, the value of this threshold depends on the full timeseries `T` (and not just one subsequence); See line below:

https://github.com/TDAmeritrade/stumpy/blob/203aa3f7721509d1139e451c43bb3aaccc63d713/tests/naive.py#L2088

A subequence `S` (in the timeseries `T`) has a certain value for its standard deviation; however, the threshold `stddev_threshold`, in the line shown above, DEPENDS on the stddev of ALL subseqs in `T`. So, when I update `T`, the threshold might change. In other words, `S` might be constant in `T`, but may not be constant in `np.append(T[1:], new_value)`. 


Now, let's see what is happening in `floss`:

```
# when `floss` object is initialized, we compute `self._T_subseq_isconstant`
self._T_subseq_isconstant = core.rolling_isconstant(self._T, self._m, self._T_subseq_isconstant_func)

# Then when we call `update()`, we have:
self._T_subseq_isconstant[:-1] = self._T_subseq_isconstant[1:]
self._T_subseq_isconstant[-1] = core.rolling_isconstant(self._T[-self._m :], self._m, self._T_subseq_isconstant_func)
```
Here, we shift the array so as to avoid computing the full array everytime. Note that the last line applies the function `self._T_subseq_isconstant_func` on ONLY ONE SUBSEQUENCE as opposed to the whole time series. Also note that the program DOES NOT MEMORIZE / LEARN the threshold `stddev_threshold` and instead it just compute it whenever the function `self._T_subseq_isconstant_func` is called. So, a different result can be obtained if, instead of shifting `self._T_subseq_isconstant`, we compute it for the whole data (and not just the last subsequence). In other words, the outcome of
```
self._T_subseq_isconstant = core.rolling_isconstant(self._T, self._m, self._T_subseq_isconstant_func)
```
can be different. Again, just to be clear, this is only because of the way the function `isconstant_func_stddev_threshold` is designed, where the stddev threshold depends on the sliding_stddev of the subsequences in the first argument of this function.

---

Now, if we look at the test function `test_floss.py`, we can see the following line:
```
ref_T_subseq_isconstant = custom_func(ref_T, m)
```
Here, we compute "subseq_isconstant " for the full array `ref_T`. (`ref_T` is a shifted version of data[:n], see the for loop ` for i, ref_T in enumerate(rolling_Ts):` in `the test_floss.py`) 

To wrap up, if I want to use the function `isconstant_func_stddev_threshold(T, m, ...)` then, I need to avoid shifiting `self._T_subseq_isconstant `, and instead just recompute it for the array as follows:
```
# in .update()
T_subseq_isconstant = core.rolling_isconstant(self._T, self._m, func)
```
note that `self._T` is the updated version of `T`. Also, I do not need to store `T_subseq_isconstant ` as attribute since we need to recompute the full array everytime. The downside is we may increase the computing time of `.update()`. Also, I am not sure even if it makes any sense to have a function to say a subsequence `S` is constant in `T` but then it says it is not constant in the shifted version of `T`. So, maybe it may not be meaningful to have a funciton like `isconstant_func_stddev_threshold`, where its output regarding a particular subsequence can be different between `T` and `np.append(T[1:], new_value)`.

seanlaw replied:
Thanks for the explanation. So, the issue seems to be with `isconstant_func_stddev_threshold` in that it requires knowledge of the FULL time series in order to compute the stddev threshold. Here are some thoughts/ideas:

Move `stddev_threshold` into the list of parameters of `isconstant_func_stddev_threshold` so that it looks like:
```
def isconstant_func_stddev_threshold(a, w, quantile_threshold=0, stddev_threshold=None):
    sliding_stddev = rolling_nanstd(a, w)
    if stddev_threshold is not None:
        return sliding_stddev <= stddev_threshold
    elif quantile_threshold == 0:  # pragma: no cover
        return sliding_stddev == 0
    else:
        stddev_threshold = np.quantile(sliding_stddev, quantile_threshold)
        return sliding_stddev <= stddev_threshold
```

Then, instead of computing `stddev_threshold` for the full time series, I would simply compute `stddev_threshold` ONCE AND ONLY ONCE for the initial `T` chunk and then, for each new data point and corresponding new subsequence, use the same pre-computed `stddev_threshold` from the first chunk. In other words, you make an assumption that all future, unseen data will have the same `stddev_threshold`.

To me, this is somewhat similar to ML feature engineering when you use [MinMaxScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html) to force your training data to have values between `0` and `1`. However, it is completely possible that your test set may have values that are larger than your `max` training data and therefore causes you to have test values that are larger than `1`. It's an assumption and I think it's okay especially for unit testing.

NimaSarajpoor replied:
> Then, instead of computing `stddev_threshold` for the full time series, I would simply compute `stddev_threshold` ONCE AND ONLY ONCE for the initial `T` chunk and then, for each new data point and corresponding new subsequence, use the same pre-computed `stddev_threshold` from the first chunk

Cool! I will revise it.

> To me, this is somewhat similar to ML feature engineering when you use [MinMaxScaler](https://scikit-learn.org/stable/modules/generated/sklearn.preprocessing.MinMaxScaler.html) to force your training data to have values between `0` and `1`. However, it is completely possible that your test set may have values that are larger than your `max` training data and therefore causes you to have test values that are larger than `1`. It's an assumption and I think it's okay especially for unit testing.

Exactly! We assume the initial `T` chunk represents the distribution of stddev for FULL data.

NimaSarajpoor commented:
@seanlaw 
Could you please checkout the error? and re-run the checks if you think that should resolve the issue.

NimaSarajpoor commented on stumpy/floss.py:
```diff
@@ -514,6 +551,17 @@ def __init__(
             )
+            nn_subseq_isconstant = self._T_subseq_isconstant[
+                self._mp[:, 3].astype(np.int64)
+            ]
```
As discussed in #808, we may add a comment to inform users of the fact that some of the arrays in `right_nn` and some booean values in `nn_subseq_isconstant` are meaningless as `self._mp[:, 3]` contains `-1`, which is not the start index of  a subsequence.  Howver, the postprocessing at the end of this "if/else block" takes care of this matter.

@seanlaw : Any suggestion?

seanlaw replied:
Is it to inform users or to inform contributors? I think we should be consistent and follow what we said earlier by adding a comment/reminder

NimaSarajpoor replied:
@seanlaw 
Oops! I meant `contributors`. I am going to use what you said in #808 

```
"Technically, a negative index value has no meaning (i.e., it is an invalid index). "
"However, in numpy, a negative value corresponds to the last position of an array "
"and so we can't "trust" that negative index value anymore and we need to replace it "
"with something else that is not -1."
```

I may do some minor changes if need. But, the main message will follow yours.

NimaSarajpoor commented on stumpy/floss.py:
```diff
@@ -514,12 +551,32 @@ def __init__(
+        # lines, if subsequence `S_i` has no valid right nearest neigbor
+        # (i.e. self._mp[i, 3]==3), then `self._mp[i, 3]` will be set to `i`. This
+        # modification results in having distance 0 (instead of np.inf) between
+        # `S_i` and its nearest neighbor. This can be resolveed by applying a
```
```suggestion
        # `S_i` and its invalid nearest neighbor. This can be resolved by applying a
```

NimaSarajpoor commented on stumpy/floss.py:
```diff
@@ -514,12 +551,32 @@ def __init__(
                 core.rolling_window(self._T, self._m) - right_nn,
                 axis=1,
                 ord=self._p,
             )
```
Below I have added the comment based on what you mentioned before. I expanded it a little. But now I think it may be too verbose. If you feel the same way, I can prune it.

seanlaw replied:
Yes, it is too long. How about:

> Note that a negative matrix profile index (e.g. `-1`) represents a null index. However, in numpy, negative indices are interpreted as counting from the end of the array. To resolve this, we temporarily replace a negative index with a self index. So, if subsequence `S_i` has no valid right nearest neighbor, then `self._mp[i, 3]` will be set to `i` and its corresponding matrix profile distance is set to 0.0. This temporary resolution is later resolved (by setting them back to `-1` and `np.inf`) in a post-processing step.

seanlaw replied:
I kinda like that as it provides just enough of a reminder that is clear and concise

NimaSarajpoor replied:
> clear and (mostly) concise

Right! I will revise the comment. Thank you for the feedback :)

NimaSarajpoor commented on stumpy/floss.py:
```diff
@@ -626,6 +699,12 @@ def P_(self):
+        in this (right) matrix profile indices is replaced with the start
+        index of the corresponding subsequence.
+        Note2: The indices stored in `self.I_` reflect the starting index of
+        subsequneces with respect to the original time series.
```
I have also added two notes here that can help user better understand the information stored in `self._I`

seanlaw replied:
You don't need `Note1` and `Note2`:

> The right matrix profile indices reflect the starting index of subsequences with respect to the original time series and all invalid matrix profile indices are replaced with the start index of the corresponding subsequence (rather than with `-1`).

I can't remember if we should/ned to keep bad indices as the "self index". I think we can/should convert the index back to `-1` as a post-processing step (and then we can omit the second half of the note above). However, we should make a copy first and then hand it back to the user

NimaSarajpoor replied:
@seanlaw 
> I can't remember if we should/ned to keep bad indices as the "self index".

Me neither :) I think I wanted to do that but after realizing the concept behind it, I did not think about it again.

> I think we can/should convert the index back to `-1` as a post-processing step

As you suggested in #808, we can do:
```
I = self._mp[:, 3].astype(np.int64).copy()
I[self._mp[:, 0] == np.inf] = -1
```

I might have some thoughts regarding the `self._I` that I can share with you in the future when I finish my investigation. 

For now, I think we should be good. I will revise the code.

seanlaw replied:
I think this should say "with respect to the full time series (i.e., including all egressed data points)"

seanlaw commented on stumpy/floss.py:
```diff
@@ -486,6 +517,12 @@ def __init__(
+            self._T_subseq_isconstant = core.rolling_isconstant(
+                self._T, self._m, self._T_subseq_isconstant_func
+            )
+            self._M_T, self._Σ_T = core.compute_mean_std(self._T, self._m)
```
When we close this PR, we should also close #778  as well, right?

NimaSarajpoor replied:
Yes, we can close that issue.

seanlaw commented on stumpy/floss.py:
```diff
@@ -571,15 +629,30 @@ def update(self, t):
+                T_subseq_isconstant=self._T_subseq_isconstant,
+                Q_subseq_isconstant=self._Q_subseq_isconstant,
             )
+            D[self._Q_subseq_isconstant & self._T_subseq_isconstant] = 0
```
Maybe apply the same `mask` here as above for consistency? Or simply add the same comment as above

NimaSarajpoor replied:
I will. 

---

Now that I am looking at this again, I think we should be able to avoid this post-processing part. Because, `mass` is a public API, and we are already passing these two arguments `Q_subseq_isconstant` and `T_subseq_isconstant` to this function. So, this `mass` function itself is taking care of this. In fact, if we follow the codes, we can see we pass these two params all the way to function `_calculate_squared_distance` (in `core.py`). .

seanlaw commented on stumpy/floss.py:
```diff
@@ -626,8 +693,14 @@ def P_(self):
         """
-        return self._mp[:, 3].astype(np.int64)
+        I = self._mp[:, 3].astype(np.int64).copy()
+        I[self._mp[:, 0] == np.inf] = -1
```
I would feel more comfortable checking both condition with something like:

```
mask = np.logical_or(self._mp[:, 0] == np.inf, self._mp[:, 3] == np.arange(len(self._mp)))
I[mask] = -1
```

NimaSarajpoor replied:
I understand your concern as I think you want to be safer here. However, you may want to reconsider it as the new check, i.e `self._mp[:, 3] == np.arange(len(self._mp)`, can be only useful at the initial step where floss object is initialized. Even in that case, the main check `self._mp[:, 0] == np.inf` should be clear and less complicated. Note that with each call of method `.update`, we shift the indices to the left and add a new value `last_idx` at the end which is bigger than `len(mp)`. In other words, `elf._mp[:, 3] == np.arange(len(self._mp))` becomes useless as soon as we start using `.update()`

seanlaw replied:
> In other words, self._mp[:, 3] == np.arange(len(self._mp)) becomes useless as soon as we start using .update()

Ahh, good point! Can you please add a note here with something like:

```
# Note that we avoid comparing the right matrix profile index value with the # self index position (i.e., self._mp[:, 3] == np.arange(len(self._mp))
# because things are constantly being shifted (ingress+egress) to the left
# and so the self._mp[:, 3] == np.arange(len(self._mp)) becomes useless as # soon as we start using .update().
```

Maybe something like that?

NimaSarajpoor replied:
Sure. I will add a comment regarding this matter.

seanlaw commented on tests/naive.py:
```diff
@@ -2080,10 +2080,12 @@ def find_matches(D, excl_zone, max_distance, max_matches=None):
 
-def isconstant_func_stddev_threshold(a, w, quantile_threshold=0):
+def isconstant_func_stddev_threshold(a, w, quantile_threshold=0, stddev_threshold=None):
     sliding_stddev = rolling_nanstd(a, w)
```
call it `rolling_stddev` instead

seanlaw commented on tests/naive.py:
```diff
@@ -2080,10 +2080,12 @@ def find_matches(D, excl_zone, max_distance, max_matches=None):
-    else:
-        stddev_threshold = np.quantile(sliding_stddev, quantile_threshold)
-        return sliding_stddev <= stddev_threshold
+    if stddev_threshold is None:
```
```
if stddev_threshold is None:
    stddev_threshold = np.quantile(sliding_stddev, quantile_threshold)
    if quantile_threshold == 0:
        stddev_threshold = 0
```

I didn't love the `else` as it was a bit harder to read due to it being nested

NimaSarajpoor replied:
@seanlaw 
I totally agree with you. What do you suggest @seanlaw ? I can go with what you initially suggested and `return` an ouput `if stddev_threshold is not None` 

```
rolling_stddev = rolling_nanstd(a, w)
if stddev_threshold is not None:
     return  rolling_stddev <  stddev_threshold

# else, meaning `stddev_threshold` is None
 if quantile_threshold == 0:
       stddev_threshold = 0  # reset to 0 
else:
       stddev_threshold = np.quantile(sliding_stddev, quantile_threshold)
```

I personally think we can replace the if/else block with this:

```
# else, meaning `stddev_threshold` is None
 stddev_threshold = np.quantile(sliding_stddev, quantile_threshold)
 if quantile_threshold == 0:
       stddev_threshold = 0  # replce computed `stddev_threshold` with 0
```

seanlaw replied:
Hmm, is there a problem with what I proposed above?

```
rolling_stddev = rolling_nanstd(a, w)

if stddev_threshold is None:
    stddev_threshold = np.quantile(rolling_stddev, quantile_threshold)
    if quantile_threshold == 0:
        stddev_threshold = 0

return rolling_stddev < stddev_threshold
```

Am I overlooking some logic here? It seems reasonably clear and okay to keep it all in my head. What I also don't love is returning a value in the middle of a function rather than at the end.

NimaSarajpoor replied:
I thought you were just quoting a part of my code and you just asked me to find a way to avoid nested if. I now noticed that the code in your initial comment was actually your suggestion :) All good! 

I knew about your concern about having return in the middle of the function but I thought you may allow this time :) I will revise the code accordingly.

NimaSarajpoor commented on tests/test_floss.py:
```diff
@@ -411,3 +429,91 @@ def test_aamp_floss_inf_nan(substitute, substitution_locations):
             npt.assert_almost_equal(ref_P, comp_P)
             npt.assert_almost_equal(ref_I, comp_I)
             npt.assert_almost_equal(ref_T, comp_T)
+            npt.assert_almost_equal(ref_mp[:, 3], stream._mp[:, 3])
```
I added this test but I now think it might not be needed. what do you think @seanlaw ?

NimaSarajpoor commented:
@seanlaw 

> Aside from one comment, everything looks good on my end

I addressed your comment. I think everything should be good now. Please feel free to merge.

seanlaw commented on stumpy/floss.py:
```diff
@@ -476,6 +498,15 @@ def __init__(
         self._custom_iac = custom_iac
         self._normalize = normalize
         self._p = p
```
I think it's important to initialize some variables:

```
self._T_subseq_isconstant = None
self._M_T = None
self._Σ_T = None
```

NimaSarajpoor replied:
Fixed!

seanlaw commented:
Apologies. I just noticed one minor thing. Once you fix that then I will merge

seanlaw commented:
@NimaSarajpoor Thank you for this contribution!
